akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/WorkPullingProducerController.scala

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.internal.delivery
+
+import java.util.concurrent.ThreadLocalRandom
+
+import scala.reflect.ClassTag
+
+import akka.Done
+import akka.actor.typed.ActorRef
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.ActorContext
+import akka.actor.typed.scaladsl.Behaviors
+
+object WorkPullingProducerController {
+
+  sealed trait InternalCommand
+
+  sealed trait Command[A] extends InternalCommand
+
+  private final case class Next[A](sendNextTo: ActorRef[A], belongsTo: ActorRef[ConsumerController.Command[A]])
+      extends InternalCommand
+
+  private final case class OutState[A](
+      producerController: ActorRef[ProducerController.Command[A]],
+      sendNextTo: Option[ActorRef[A]])
+
+  private final case class State[A](
+      consumers: Map[ActorRef[ConsumerController.Command[A]], OutState[A]],
+      hasRequested: Boolean)
+
+  // TODO Now the workers have to be registered explicitly/manually.
+  // We could support automatic registration via Receptionist, similar to how routers work.
+
+  final case class RegisterWorker[A](
+      consumerController: ActorRef[ConsumerController.Command[A]],
+      replyTo: ActorRef[Done])
+      extends Command[A]
+
+  private final case class Msg[A](msg: A) extends InternalCommand
+
+  def apply[A: ClassTag, RequestNext](
+      producerId: String,
+      requestNextFactory: ActorRef[A] ⇒ RequestNext,
+      producer: ActorRef[RequestNext]): Behavior[Command[A]] = {
+    Behaviors
+      .setup[InternalCommand] { context =>
+        val msgAdapter: ActorRef[A] = context.messageAdapter(msg ⇒ Msg(msg))
+        val requestNext = requestNextFactory(msgAdapter)
+        new WorkPullingProducerController(context, producerId, requestNext, producer).active(
+          State(Map.empty, hasRequested = false))
+      }
+      .narrow
+  }
```

Comment:
This is very cool, kind of pre-baking much of the distributed workers sample.

Reply:
yup, was fun to write